sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,912 @@
+//! @brief Solana Rust-based BigNumber for bpf-programs
+
+use borsh::{BorshDeserialize, BorshSchema, BorshSerialize};
+use std::fmt;
+
+#[repr(C)]
+#[derive(
+    BorshSerialize,
+    BorshDeserialize,
+    BorshSchema,
+    Clone,
+    Eq,
+    PartialEq,
+    Ord,
+    PartialOrd,
+    Hash,
+    AbiExample,
+)]
+
+/// BigNumber supports integers greater than
+/// 64 bits for BPF programs
+pub struct BigNumber {
+    value: Vec<u8>,
+    is_negative: bool,
+}
+
+impl BigNumber {
+    /// Returns a BigNumber with initial value of 0
+    pub fn new() -> Self {
+        Self {
+            is_negative: false,
+            value: vec![0u8],
+        }
+    }
+
+    /// Returns the size, in bytes, of BigNum. Typically used in
+    /// assessing buffer size needed to fetch the bytes for serialization, etc.
+    pub fn size_in_bytes(&self) -> usize {
+        self.value.len()
+    }
+
+    /// Returns a BigNumber with initial value set to a u32 value
+    pub fn from_u32(val: u32) -> Result<Self, u64> {
```

Comment:
Instead of `u64` how about returning a more meaningful error (an error enum).  This would also allow for using the `?` operator rather then the nested matches which all return `Err(1)`

Reply:
Yes, that was one of the items for reworking I mentioned.

Reply:
As per previous discussions the BigNumber methods only return positive results, all other errors halt the program for an early exit